apps/insights/src/components/PriceFeed/Chart/chart.tsx

```diff
@@ -445,6 +462,119 @@ const useChartColors = (
   }, [resolvedTheme, chartRef, chartContainerRef]);
 };
 
+const useChartHoverCard = (
+  chartRef: RefObject<ChartRefContents | undefined>,
+  chartContainerRef: RefObject<HTMLDivElement | null>,
+  priceFormatter: ReturnType<typeof usePriceFormatter>,
+) => {
+  const dateFormatter = useDateFormatter({
+    year: "2-digit",
+    month: "short",
+    day: "numeric",
+    hour: "2-digit",
+    minute: "2-digit",
+    second: "2-digit",
+    hour12: false,
+    timeZone: "UTC",
+  });
+
+  const hoverCardRef = useRef<HTMLDivElement | null>(null);
+  const hoverCard = hoverCardRef.current;
+  const [hoverCardData, setHoverCardData] = useState<
+    | (ChartHoverCardProps & {
+        style: {
+          left: string;
+          top: string;
+        };
+      })
+    | undefined
+  >(undefined);
+
+  useEffect(() => {
+    const chartData = chartRef.current;
+    const chartElem = chartContainerRef.current;
+
+    if (!chartData || !chartElem || !hoverCard) {
+      return;
+    }
+
+    const { chart, price, confidenceHigh, confidenceLow } = chartData;
+
+    const handleCrosshairMove: Parameters<
+      typeof chart.subscribeCrosshairMove
+    >[0] = (param) => {
+      const priceData = param.seriesData.get(price);
+      const confidenceHighData = param.seriesData.get(confidenceHigh);
+      const confidenceLowData = param.seriesData.get(confidenceLow);
+
+      const hasPrice = priceData && "value" in priceData;
+      const hasTime = param.time !== undefined;
+
+      if (param.point === undefined || !hasPrice || !hasTime) {
+        setHoverCardData(undefined);
+        return;
+      }
+
+      const priceValue = priceData.value;
+      const timestampValue = new Date(Number(param.time) * 1000);
+      const formattedTimestamp = dateFormatter.format(timestampValue);
+      const formattedPrice = priceFormatter.format(priceValue);
+
+      let formattedConfidence = "N/A";
+      if (
+        confidenceHighData &&
+        "value" in confidenceHighData &&
+        confidenceLowData &&
+        "value" in confidenceLowData
+      ) {
+        formattedConfidence = `±${priceFormatter.format(confidenceHighData.value - priceValue)}`;
+      }
+
+      const hoverCardRect = hoverCard.getBoundingClientRect();
```

Comment:
if you can, I would recommend setting up resize observers for the hoverCard and chartElem. calling `getBoundingClientRect()` can cause full page reflows, which, given how often this event might occur (while the user is moving the mouse), could be computationally expensive and could impact performance on lower-end hardware.

we could use resize observers for each of these elems and do `setState()` for the required rect values instead to avoid this